realm/realm-library/src/main/java/io/realm/BaseRealm.java

```diff
@@ -70,7 +72,10 @@
     // thread pool for all async operations (Query & transaction)
     static final RealmThreadPoolExecutor asyncQueryExecutor = RealmThreadPoolExecutor.getInstance();
 
-    protected final List<WeakReference<RealmChangeListener>> changeListeners =
+    protected final Set<RealmChangeListener> changeListeners =
```

Comment:
User shouldn't worry about the order. 
I'll revert to `CopyOnWriteArrayList` though since `CopyOnWriteArraySet` is backed by `CopyOnWriteArrayList` anyway, then we can call `addIfAbsent` to avoid iterating all elements in `addChangeListener`